Content.Server/Mousetrap/MousetrapSystem.cs

```diff
@@ -0,0 +1,98 @@
+using Content.Shared.Damage;
+using Content.Shared.Interaction;
+using Content.Shared.Interaction.Events;
+using Content.Shared.Inventory;
+using Content.Shared.Mousetrap;
+using Content.Shared.StepTrigger;
+using Robust.Shared.Audio;
+using Robust.Shared.Player;
+
+namespace Content.Server.Mousetrap;
+
+public sealed class MousetrapSystem : EntitySystem
+{
+    [Dependency] private readonly DamageableSystem _damageableSystem = default!;
+    [Dependency] private readonly InventorySystem _inventorySystem = default!;
+
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<MousetrapComponent, UseInHandEvent>(OnUseInHand);
+        SubscribeLocalEvent<MousetrapComponent, StepTriggerAttemptEvent>(OnStepTriggerAttempt);
+        SubscribeLocalEvent<MousetrapComponent, StepTriggeredEvent>(OnStepTrigger);
+    }
+
+    private void Trigger(EntityUid uid, MousetrapComponent? component = null)
+    {
+        if (!Resolve(uid, ref component))
+        {
+            return;
+        }
+
+        SoundSystem.Play(Filter.Pvs(uid), component.SoundOnActivate.GetSound());
+
+        component.IsActive = false;
+    }
+
+    private void OnUseInHand(EntityUid uid, MousetrapComponent component, UseInHandEvent args)
+    {
+        component.IsActive = !component.IsActive;
+
+        UpdateVisuals(uid);
+    }
+
+    private void OnStepTriggerAttempt(EntityUid uid, MousetrapComponent component, ref StepTriggerAttemptEvent args)
+    {
+        if (!component.IsActive)
+        {
+            return;
+        }
+
+        var prototype = Prototype(args.Tripper);
+
+        if (prototype != null
+            && component.SpecialDamageEntities.TryGetValue(prototype.ID, out var damage))
+        {
+            _damageableSystem.TryChangeDamage(args.Tripper, damage, true);
+            Trigger(uid, component);
+            return;
+        }
+
+        foreach (var slot in component.IgnoreDamageIfSlotFilled)
+        {
+            if (!_inventorySystem.TryGetSlotContainer(args.Tripper, slot, out var container, out _))
+            {
+                continue;
+            }
+
+            // Yes, this also means that wearing slippers won't
+            // hurt the entity.
+            if (container.ContainedEntity != null)
+            {
+                Trigger(uid, component);
+                break;
+            }
+        }
+
+        args.Continue = component.IsActive;
+    }
+
+    private void OnStepTrigger(EntityUid uid, MousetrapComponent component, ref StepTriggeredEvent args)
+    {
+        Trigger(uid, component);
+
+
+
+        UpdateVisuals(uid);
+    }
```

Comment:
why does the actual step trigger also if the attempt event does the same thing?

Reply:
this will only occur if the mousetrap is still active, ala the continue field in the event